include/picongpu/particles/boundary/Absorbing.hpp

```diff
@@ -0,0 +1,145 @@
+/* Copyright 2021 Sergei Bastrakov
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+
+#include "picongpu/particles/boundary/ApplyImpl.hpp"
+#include "picongpu/particles/boundary/Kind.hpp"
+#include "picongpu/particles/manipulators/unary/FreeTotalCellOffset.hpp"
+
+#include <pmacc/Environment.hpp>
+
+#include <cstdint>
+
+
+namespace picongpu
+{
+    namespace particles
+    {
+        namespace boundary
+        {
+            //! Functor to apply absorbing boundary to particle species
+            template<>
+            struct ApplyImpl<Kind::Absorbing>
+            {
+                //! Functor to be applied to all particles in the active area
+                class AbsorbParticleIfOutside
+                {
+                public:
+                    //! Some name is required
+                    static constexpr char const* name = "absorbParticleIfOutside";
+
+                    //! Construct an instance, the parameters must be passed via staticParameters()
+                    AbsorbParticleIfOutside() : parameters(staticParameters())
```

Comment:
@psychocoderHPC this is quite awkward (also see that `staticParameters()` function). Do you know a better way? I understand the "global" data can be queried just inside this constructor via `Environment`. However, I also need to pass data that is only available in `ApplyImpl<Kind::Absorbing>::operator()` and nowhere else, and that pushed me to this ugly stuff.

Reply:
unfortunately, I do not know a better way :-(

Reply:
I've now refactored it into a generic basic functor (last commit). So the ugly way is the same, but at least it's separated, documented and reusable.